backendoori/ootw-backend
# ✒️ Must Have 기능 구현 및 리뷰

## 백엔두리 팀 코멘트

아래는 구현된 Must Have 기능을 정리한 내용입니다! 편하실 때 피드백 해주시면 감사하겠습니다!

 [API 명세서는 해당 링크에서 확인 가능합니다!](https://docs.google.com/spreadsheets/d/1RtuZxomfPegPjpzhH1FJf4209MCSPqjVAthiHskIA4w/edit?usp=sharing)

## Must Have 기능 정리 및 진행도

### 1. 인증 기능

- 1-1. 회원 가입 기능
    - [x]  기능 구현 및 작동 확인
    - [x]  테스트 코드 작성
    - [ ]  유효성 검사
- 1-2. 로그인 기능
    - [x]  기능 구현 및 작동 확인
    - [x]  테스트 코드 작성
    - [ ]  유효성 검사

### 2. 아바타 관련 기능

- 2-1. 아바타 이미지 생성 기능
    - [x]  이미지 업로드를 위한 storage 컨테이너 구축(MinIO)
    - [x]  이미지 업로드 서비스 기능 구축
    - [x]  이미지 업로드 정상 작동 확인
    - [x]  테스트 코드
    - [ ]  유효성 검사 **필요**
    - [ ]  테스트 코드 **고도화**

### 3. 커뮤니티 관련 기능

- 3-1. 게시글 생성 기능
    - [x]  기능 구현 및 작동 확인
    - [x]  테스트 코드 작성
    - [x]  유효성 검사
- 3-2. 게시글 단건 조회 기능
    - [x]  기능 구현 및 작동 확인
    - [x]  테스트 코드 작성
    - [x]  유효성 검사
- 3-3. 게시글 목록 조회 기능
    - [x]  기능 구현 및 작동 확인
    - [x]  테스트 코드 작성
    - [x]  유효성 검사
- 3-4. 현재 날씨 불러오기 기능
    - [x]  기능 구현 및 작동 확인
    - [ ]  테스트 코드 작성 **필요**
    - [ ]  유효성 검사 **필요**

### 5. **멘토님 미션**

- 5-1. JACOCO
    - [x]  Gradle Task 추가
        - gradle test가 실행될 때 마다 다음의 task가 자동적으로 실행
        - jacocoTestReport
        - jacocoTestCoverageVerification
    - [x]  Test 커버리지 기준 설정
        - `Class`는 **코드 라인의 70%가 COVER** 되어야 함
        - `Class` 내부의 **Branch**(ex. `if`, `switch`)는 **코드 라인의 70%가 COVER** 되어야 함
        - `Method`는 **코드 라인이 20줄 이내** 여야 함
    - [x]  Test 커버리지 report 생성
- 5-2. CI
    - [x]  github-action을 이용한 CI 적용
        - pull reque 시 CI를 통한 검증
        - build 단계에서 check style과 jacoco 검증
        - 실제 docker-compose를 띄워 로컬에서 작업과 동일한 환경에서 진행
- 5-3. CHECK STYLE
    - [x]  google checkstyle을 기반으로 만든 checkstyle 커스텀 파일 적용
        - 만들어진 checkstyle를 기반으로 intellij code style 커스텀

## 프로젝트 구조도 설명

```bash
├─avatar  🚀아바타 관련 기능
│  ├─controller
│  ├─domain
│  ├─dto
│  ├─repository
│  └─service
├─common
│  └─image  🚀이미지 관련 기능
├─config
├─exception
├─post  🚀커뮤니티 관련 기능
│  ├─controller
│  ├─domain
│  │  └─weather
│  ├─dto
│  ├─repository
│  └─service
├─security  
│  └─jwt
├─user  🚀사용자 인증 관련 기능
│  ├─controller
│  ├─domain
│  ├─dto
│  ├─repository
│  └─service
└─weather  🚀날씨 관련 기능
    ├─controller
    ├─domain
    ├─dto
    ├─service
    └─util
```

## Reference site

- [꾸온꾸](https://korean.visitkorea.or.kr/detail/rem_detail.do?cotid=14a13909-134a-4a0e-a581-fc1ace95d925)
